Sources/bip32-ed25519-swift/bip32-ed25519-swift.swift

```diff
@@ -242,5 +256,89 @@ func deriveChildNodePrivate(extendedKey: Data, index: UInt32) -> Data {
     public func verifyWithPublicKey(signature: Data, message: Data, publicKey: Data) -> Bool {
         return SodiumHelper.cryptoSignVerifyDetached(signature, message,publicKey)
     }
+
+    func hasAlgorandTags(data: Data) -> Bool {
+        // Prefixes taken from go-algorand node software code
+        // https://github.com/algorand/go-algorand/blob/master/protocol/hash.go
+
+        let prefixes = ["appID", "arc", "aB", "aD", "aO", "aP", "aS", "AS", "BH", "B256", "BR", "CR", "GE", "KP", "MA", "MB", "MX", "NIC", "NIR", "NIV", "NPR", "OT1", "OT2", "PF", "PL", "Program", "ProgData", "PS", "PK", "SD", "SpecialAddr", "STIB", "spc", "spm", "spp", "sps", "spv", "TE", "TG", "TL", "TX", "VO"]
+        let prefixBytes = prefixes.map { $0.data(using: .ascii)! }
+        return prefixBytes.contains { data.starts(with: $0) }
+    }
+
+    public func validateData(data: Data, metadata: SignMetadata) throws -> Bool {
+
+        if hasAlgorandTags(data: data) {
+            return false
+        }
+
+        // Validate data based on the schema
+        var rawData: Data
+        switch metadata.encoding {
+            case .base64:
+                guard let string = String(data: data, encoding: .utf8),
+                    let base64Data = Data(base64Encoded: string) else {
+                    return false
+                }
+                rawData = base64Data
+            case .msgpack:
+                do {
+                    rawData = try JSONSerialization.data(withJSONObject: messagePackValueToSwift(try MessagePack.unpack(data).value), options: [])
+                } catch {
+                    return false
+                }
+            case .none:
+                rawData = data
+        }
+
+        do {
+            let valid = try JSONSchema.validate(try JSONSerialization.jsonObject(with: rawData, options: []) as! [String: Any], schema: metadata.schema)
+            return valid.valid
+        } catch {
+            return false
+        }
+    }
+
+    public func signData(context: KeyContext, account: UInt32, change: UInt32, keyIndex: UInt32, data: Data, metadata: SignMetadata) throws -> Data {
+        let valid = try validateData(data: data, metadata: metadata)
+
+        if !valid{
+            throw DataValidationException(message: "Data is not valid")
+        }
+
+        let bip44Path: [UInt32] = getBIP44PathFromContext(context: context, account: account, change: change, keyIndex: keyIndex)
+        return rawSign(bip44Path: bip44Path, message: data)
+    }
+
+    func messagePackValueToSwift(_ value: MessagePackValue) -> Any {
```

Comment:
why do we need this function? can you add a comment explaining?

Reply:
We need a way to map from Message Pack to Swift objects which can be turned into JSON. The JSON can then be checked against our JSON schema validator.

I'll add a comment.